service/worker/scanner/history/scavenger_test.go

```diff
@@ -390,79 +427,219 @@ func (s *ScavengerTestSuite) TestMixesTwoPages() {
+			ExecutionInfo: &persistencepb.WorkflowExecutionInfo{
+				LastUpdateTime: timestamp.TimePtr(time.Now()),
+			},
+		},
+	}
+	s.mockRegistry.EXPECT().GetNamespaceByID(gomock.Any()).Return(mockedNamespace, nil).AnyTimes()
+	s.mockHistoryClient.EXPECT().DescribeMutableState(gomock.Any(), &historyservice.DescribeMutableStateRequest{
 		NamespaceId: "namespaceID3",
 		Execution: &commonpb.WorkflowExecution{
 			WorkflowId: "workflowID3",
 			RunId:      "runID3",
 		},
 	}).Return(nil, serviceerror.NewNotFound(""))
 
-	client.EXPECT().DescribeMutableState(gomock.Any(), &historyservice.DescribeMutableStateRequest{
+	s.mockHistoryClient.EXPECT().DescribeMutableState(gomock.Any(), &historyservice.DescribeMutableStateRequest{
 		NamespaceId: "namespaceID4",
 		Execution: &commonpb.WorkflowExecution{
 			WorkflowId: "workflowID4",
 			RunId:      "runID4",
 		},
 	}).Return(nil, serviceerror.NewNotFound(""))
-	client.EXPECT().DescribeMutableState(gomock.Any(), &historyservice.DescribeMutableStateRequest{
+	s.mockHistoryClient.EXPECT().DescribeMutableState(gomock.Any(), &historyservice.DescribeMutableStateRequest{
 		NamespaceId: "namespaceID5",
 		Execution: &commonpb.WorkflowExecution{
 			WorkflowId: "workflowID5",
 			RunId:      "runID5",
 		},
-	}).Return(nil, nil)
+	}).Return(ms, nil)
 
 	branchToken3, err := persistence.NewHistoryBranchToken(treeID3, branchID3)
 	s.Nil(err)
-	db.EXPECT().DeleteHistoryBranch(gomock.Any(), &persistence.DeleteHistoryBranchRequest{
+	s.mockExecutionManager.EXPECT().DeleteHistoryBranch(gomock.Any(), &persistence.DeleteHistoryBranchRequest{
 		BranchToken: branchToken3,
 		ShardID:     common.WorkflowIDToHistoryShard("namespaceID3", "workflowID3", s.numShards),
 	}).Return(nil)
 
 	branchToken4, err := persistence.NewHistoryBranchToken(treeID4, branchID4)
 	s.Nil(err)
-	db.EXPECT().DeleteHistoryBranch(gomock.Any(), &persistence.DeleteHistoryBranchRequest{
+	s.mockExecutionManager.EXPECT().DeleteHistoryBranch(gomock.Any(), &persistence.DeleteHistoryBranchRequest{
 		BranchToken: branchToken4,
 		ShardID:     common.WorkflowIDToHistoryShard("namespaceID4", "workflowID4", s.numShards),
 	}).Return(fmt.Errorf("failed to delete history"))
 
-	hbd, err := scvgr.Run(context.Background())
+	hbd, err := s.scavenger.Run(context.Background())
 	s.Nil(err)
 	s.Equal(1, hbd.SkipCount)
 	s.Equal(2, hbd.SuccessCount)
 	s.Equal(2, hbd.ErrorCount)
 	s.Equal(2, hbd.CurrentPage)
 	s.Equal(0, len(hbd.NextPageToken))
 }
+
+func (s *ScavengerTestSuite) TestDeleteWorkflowAfterRetention() {
+	s.mockExecutionManager.EXPECT().GetAllHistoryTreeBranches(gomock.Any(), &persistence.GetAllHistoryTreeBranchesRequest{
+		PageSize: pageSize,
+	}).Return(&persistence.GetAllHistoryTreeBranchesResponse{
+		NextPageToken: []byte("page1"),
+		Branches: []persistence.HistoryBranchDetail{
+			{
+				BranchToken: s.toBranchToken("treeID1", "branchID1"),
+				ForkTime:    timestamp.TimeNowPtrUtcAddDuration(-s.scavenger.historyDataMinAge() * 2),
```

Comment:
Shouldn't these be based off of the retention and not the `historyDataMinAge`?

Reply:
Updated.